File: app/src/main/java/com/quranapp/android/utils/sp/SPReader.java
package com.quranapp.android.utils.sp;

import static com.quranapp.android.readerhandler.ReaderParams.READER_STYLE_DEFAULT;
import static com.quranapp.android.utils.reader.ScriptUtils.KEY_SCRIPT;
import static com.quranapp.android.utils.reader.ScriptUtils.SCRIPT_DEFAULT;
import static com.quranapp.android.utils.reader.TextSizeUtils.KEY_TEXT_SIZE_MULT_ARABIC;
import static com.quranapp.android.utils.reader.TextSizeUtils.KEY_TEXT_SIZE_MULT_TRANSL;
import static com.quranapp.android.utils.reader.TextSizeUtils.TEXT_SIZE_MULT_AR_DEFAULT;
import static com.quranapp.android.utils.reader.TextSizeUtils.TEXT_SIZE_MULT_TRANS_DEFAULT;
import static com.quranapp.android.utils.reader.TranslUtils.KEY_TRANSLATIONS;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.KEY_RECITATION_CONTINUE_CHAPTER;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.KEY_RECITATION_RECITER;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.KEY_RECITATION_REPEAT;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.KEY_RECITATION_VERSE_SYNC;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.RECITATION_DEFAULT_CONTINUE_CHAPTER;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.RECITATION_DEFAULT_REPEAT;
import static com.quranapp.android.utils.reader.recitation.RecitationUtils.RECITATION_DEFAULT_VERSE_SYNC;
import static com.quranapp.android.utils.univ.Keys.READER_KEY_READER_STYLE;

import android.content.Context;
import android.content.SharedPreferences;

import com.quranapp.android.utils.reader.TranslUtils;

import java.util.HashSet;
import java.util.Set;

/**
 * SharedPreferences utility class for Reader
 */
public abstract class SPReader {
    public static final String SP_TEXT_STYLE = "sp_reader_text";
    public static final String SP_TRANSL = "sp_reader_translations";
    public static final String SP_RECITATION_OPTIONS = "sp_reader_recitation_options";
    public static final String SP_SCRIPT = "sp_reader_script";
    public static final String SP_READER_STYLE = "sp_reader_style";

    public static float getSavedTextSizeMultArabic(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_TEXT_STYLE, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_TEXT_SIZE_MULT_ARABIC)) {
            setSavedTextSizeMultArabic(context, TEXT_SIZE_MULT_AR_DEFAULT);
        }

        return sp.getFloat(KEY_TEXT_SIZE_MULT_ARABIC, TEXT_SIZE_MULT_AR_DEFAULT);
    }

    public static void setSavedTextSizeMultArabic(Context context, float sizeMult) {
        SharedPreferences sp = context.getSharedPreferences(SP_TEXT_STYLE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putFloat(KEY_TEXT_SIZE_MULT_ARABIC, sizeMult);
        editor.apply();
    }

    public static float getSavedTextSizeMultTransl(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_TEXT_STYLE, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_TEXT_SIZE_MULT_TRANSL)) {
            setSavedTextSizeMultTransl(context, TEXT_SIZE_MULT_TRANS_DEFAULT);
        }

        return sp.getFloat(KEY_TEXT_SIZE_MULT_TRANSL, TEXT_SIZE_MULT_TRANS_DEFAULT);
    }

    public static void setSavedTextSizeMultTransl(Context context, float sizeMult) {
        SharedPreferences sp = context.getSharedPreferences(SP_TEXT_STYLE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putFloat(KEY_TEXT_SIZE_MULT_TRANSL, sizeMult);
        editor.apply();
    }

    public static Set<String> getSavedTranslations(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_TRANSL, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_TRANSLATIONS)) {
            SharedPreferences.Editor editor = sp.edit();
            editor.putStringSet(KEY_TRANSLATIONS, TranslUtils.defaultTranslationSlugs());
            editor.apply();
        }

        if (sp.contains(KEY_TRANSLATIONS)) {
            return new HashSet<>(sp.getStringSet(KEY_TRANSLATIONS, new HashSet<>()));
        }

        return new HashSet<>();
    }

    public static void setSavedTranslations(Context context, Set<String> translSlugsSet) {
        SharedPreferences sp = context.getSharedPreferences(SP_TRANSL, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putStringSet(KEY_TRANSLATIONS, new HashSet<>(translSlugsSet));
        editor.apply();
    }

    public static Set<String> addToSavedTranslations(Context context, String translSlug) {
        Set<String> savedTranslations = new HashSet<>(getSavedTranslations(context));
        savedTranslations.add(translSlug);

        SharedPreferences sp = context.getSharedPreferences(SP_TRANSL, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putStringSet(KEY_TRANSLATIONS, savedTranslations);
        editor.apply();

        return savedTranslations;
    }

    public static Set<String> removeFromSavedTranslations(Context context, String translSlug) {
        Set<String> savedTranslations = new HashSet<>(getSavedTranslations(context));
        savedTranslations.remove(translSlug);

        SharedPreferences sp = context.getSharedPreferences(SP_TRANSL, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putStringSet(KEY_TRANSLATIONS, savedTranslations);
        editor.apply();

        return savedTranslations;
    }

    public static String getSavedRecitationSlug(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);
        return sp.getString(KEY_RECITATION_RECITER, "");
    }

    public static void setSavedRecitationSlug(Context context, String recitation) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(KEY_RECITATION_RECITER, recitation);
        editor.apply();
    }

    public static boolean getRecitationRepeatVerse(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_RECITATION_REPEAT)) {
            setRecitationRepeatVerse(context, RECITATION_DEFAULT_REPEAT);
        }

        return sp.getBoolean(KEY_RECITATION_REPEAT, RECITATION_DEFAULT_REPEAT);
    }

    public static void setRecitationRepeatVerse(Context context, boolean repeatVerse) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean(KEY_RECITATION_REPEAT, repeatVerse);
        editor.apply();
    }

    public static boolean getRecitationContinueChapter(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_RECITATION_CONTINUE_CHAPTER)) {
            setRecitationContinueChapter(context, RECITATION_DEFAULT_CONTINUE_CHAPTER);
        }

        return sp.getBoolean(KEY_RECITATION_CONTINUE_CHAPTER, RECITATION_DEFAULT_CONTINUE_CHAPTER);
    }

    public static void setRecitationContinueChapter(Context context, boolean continueChapter) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean(KEY_RECITATION_CONTINUE_CHAPTER, continueChapter);
        editor.apply();
    }

    public static boolean getRecitationScrollSync(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_RECITATION_VERSE_SYNC)) {
            setRecitationVerseSync(context, RECITATION_DEFAULT_VERSE_SYNC);
        }

        return sp.getBoolean(KEY_RECITATION_VERSE_SYNC, RECITATION_DEFAULT_VERSE_SYNC);
    }

    public static void setRecitationVerseSync(Context context, boolean sync) {
        SharedPreferences sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putBoolean(KEY_RECITATION_VERSE_SYNC, sync);
        editor.apply();
    }

    public static String getSavedScript(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_SCRIPT, Context.MODE_PRIVATE);

        if (!sp.contains(KEY_SCRIPT)) {
            setSavedScript(context, SCRIPT_DEFAULT);
        }

        return sp.getString(KEY_SCRIPT, SCRIPT_DEFAULT);
    }

    public static void setSavedScript(Context context, String font) {
        SharedPreferences sp = context.getSharedPreferences(SP_SCRIPT, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putString(KEY_SCRIPT, font);
        editor.apply();
    }

    public static int getSavedReaderStyle(Context context) {
        SharedPreferences sp = context.getSharedPreferences(SP_READER_STYLE, Context.MODE_PRIVATE);

        if (!sp.contains(READER_KEY_READER_STYLE)) {
            setSavedReaderStyle(context, READER_STYLE_DEFAULT);
        }

        sp = context.getSharedPreferences(SP_READER_STYLE, Context.MODE_PRIVATE);
        return sp.getInt(READER_KEY_READER_STYLE, READER_STYLE_DEFAULT);
    }

    public static void setSavedReaderStyle(Context context, int readerStyle) {
        SharedPreferences sp = context.getSharedPreferences(SP_READER_STYLE, Context.MODE_PRIVATE);
        SharedPreferences.Editor editor = sp.edit();
        editor.putInt(READER_KEY_READER_STYLE, readerStyle);
        editor.apply();
    }
}

File: app/src/main/java/com/quranapp/android/utils/sharedPrefs/SPReader.kt
package com.quranapp.android.utils.sharedPrefs

import android.content.Context
import androidx.core.content.edit
import com.quranapp.android.compose.utils.preferences.ReaderPreferences
import com.quranapp.android.utils.reader.recitation.RecitationManager.setSavedRecitationSlug
import com.quranapp.android.utils.reader.recitation.RecitationManager.setSavedRecitationTranslationSlug
import com.quranapp.android.utils.reader.recitation.RecitationUtils
import kotlinx.coroutines.runBlocking

/**
 * SharedPreferences utility class for Reader
 */
@Deprecated("Use ReaderPreferences (DataStore)")
object SPReader {
    private const val SP_RECITATION_OPTIONS: String = "sp_reader_recitation_options"

    @JvmStatic
    fun getArabicTextEnabled(context: Context): Boolean {
        return ReaderPreferences.getArabicTextEnabled()
    }

    @JvmStatic
    fun setArabicTextEnabled(context: Context, enabled: Boolean) {
        runBlocking { ReaderPreferences.setArabicTextEnabled(enabled) }
    }

    @JvmStatic
    fun getAutoScrollSpeed(context: Context): Float {
        return ReaderPreferences.getAutoScrollSpeed()
    }

    @JvmStatic
    fun setAutoScrollSpeed(context: Context, speed: Float) {
        runBlocking { ReaderPreferences.setAutoScrollSpeed(speed) }
    }

    @JvmStatic
    fun getSavedTextSizeMultArabic(context: Context): Float {
        return ReaderPreferences.getArabicTextSizeMultiplier()
    }

    @JvmStatic
    fun setSavedTextSizeMultArabic(context: Context, sizeMult: Float) {
        runBlocking { ReaderPreferences.setArabicTextSizeMultiplier(sizeMult) }
    }

    @JvmStatic
    fun getSavedTextSizeMultTransl(context: Context): Float {
        return ReaderPreferences.getTranslationTextSizeMultiplier()
    }

    @JvmStatic
    fun setSavedTextSizeMultTransl(context: Context, sizeMult: Float) {
        runBlocking { ReaderPreferences.setTranslationTextSizeMultiplier(sizeMult) }
    }

    fun getSavedTextSizeMultTafsir(context: Context): Float {
        return ReaderPreferences.getTafsirTextSizeMultiplier()
    }

    @JvmStatic
    fun setSavedTextSizeMultTafsir(context: Context, sizeMult: Float) {
        runBlocking { ReaderPreferences.setTafsirTextSizeMultiplier(sizeMult) }
    }

    @JvmStatic
    fun getSavedTranslations(context: Context): HashSet<String> {
        return ReaderPreferences.getTranslations()
    }

    @JvmStatic
    fun setSavedTranslations(context: Context, translSlugsSet: Set<String>) {
        runBlocking { ReaderPreferences.setTranslations(translSlugsSet) }
    }

    @JvmStatic
    fun getSavedRecitationSlug(context: Context): String? {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        return sp.getString(RecitationUtils.KEY_RECITATION_RECITER, null)
    }

    @JvmStatic
    fun setSavedRecitationSlug(context: Context, recitation: String) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putString(RecitationUtils.KEY_RECITATION_RECITER, recitation)
        }

        setSavedRecitationSlug(recitation)
    }

    @JvmStatic
    fun getSavedRecitationTranslationSlug(context: Context): String? {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        return sp.getString(RecitationUtils.KEY_RECITATION_TRANSLATION_RECITER, null)
    }

    @JvmStatic
    fun setSavedRecitationTranslationSlug(context: Context, slug: String) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putString(RecitationUtils.KEY_RECITATION_TRANSLATION_RECITER, slug)
        }

        setSavedRecitationTranslationSlug(slug)
    }

    fun getRecitationSpeed(context: Context): Float {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)

        if (!sp.contains(RecitationUtils.KEY_RECITATION_SPEED)) {
            setRecitationSpeed(context, RecitationUtils.RECITATION_DEFAULT_SPEED)
        }

        return sp.getFloat(
            RecitationUtils.KEY_RECITATION_SPEED,
            RecitationUtils.RECITATION_DEFAULT_SPEED
        )
    }

    fun setRecitationSpeed(context: Context, speed: Float) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putFloat(RecitationUtils.KEY_RECITATION_SPEED, speed)
        }
    }

    fun getRecitationRepeatVerse(context: Context): Boolean {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)

        if (!sp.contains(RecitationUtils.KEY_RECITATION_REPEAT)) {
            setRecitationRepeatVerse(context, RecitationUtils.RECITATION_DEFAULT_REPEAT)
        }

        return sp.getBoolean(
            RecitationUtils.KEY_RECITATION_REPEAT,
            RecitationUtils.RECITATION_DEFAULT_REPEAT
        )
    }

    fun setRecitationRepeatVerse(context: Context, repeatVerse: Boolean) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putBoolean(RecitationUtils.KEY_RECITATION_REPEAT, repeatVerse)
        }
    }

    fun getRecitationContinueChapter(context: Context): Boolean {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)

        if (!sp.contains(RecitationUtils.KEY_RECITATION_CONTINUE_CHAPTER)) {
            setRecitationContinueChapter(
                context,
                RecitationUtils.RECITATION_DEFAULT_CONTINUE_CHAPTER
            )
        }

        return sp.getBoolean(
            RecitationUtils.KEY_RECITATION_CONTINUE_CHAPTER,
            RecitationUtils.RECITATION_DEFAULT_CONTINUE_CHAPTER
        )
    }

    fun setRecitationContinueChapter(context: Context, continueChapter: Boolean) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putBoolean(RecitationUtils.KEY_RECITATION_CONTINUE_CHAPTER, continueChapter)
        }
    }

    fun getRecitationScrollSync(context: Context): Boolean {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)

        if (!sp.contains(RecitationUtils.KEY_RECITATION_SCROLL_SYNC)) {
            setRecitationScrollSync(context, RecitationUtils.RECITATION_DEFAULT_VERSE_SYNC)
        }

        return sp.getBoolean(
            RecitationUtils.KEY_RECITATION_SCROLL_SYNC,
            RecitationUtils.RECITATION_DEFAULT_VERSE_SYNC
        )
    }

    fun setRecitationScrollSync(context: Context, sync: Boolean) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putBoolean(RecitationUtils.KEY_RECITATION_SCROLL_SYNC, sync)
        }
    }

    @JvmStatic
    fun getRecitationAudioOption(context: Context): Int {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)

        if (!sp.contains(RecitationUtils.KEY_RECITATION_AUDIO_OPTION)) {
            setRecitationAudioOption(context, RecitationUtils.AUDIO_OPTION_DEFAULT)
        }

        return sp.getInt(
            RecitationUtils.KEY_RECITATION_AUDIO_OPTION,
            RecitationUtils.AUDIO_OPTION_DEFAULT
        )
    }

    fun setRecitationAudioOption(context: Context, option: Int) {
        val sp = context.getSharedPreferences(SP_RECITATION_OPTIONS, Context.MODE_PRIVATE)
        sp.edit() {
            putInt(RecitationUtils.KEY_RECITATION_AUDIO_OPTION, option)
        }
    }

    @JvmStatic
    fun getSavedScript(context: Context): String {
        return ReaderPreferences.getQuranScript()
    }

    @JvmStatic
    fun setSavedScript(context: Context, font: String?) {
        runBlocking { ReaderPreferences.setQuranScript(font) }
    }

    @JvmStatic
    fun getSavedReaderStyle(context: Context): Int {
        return ReaderPreferences.getReaderStyle()
    }

    @JvmStatic
    fun setSavedReaderStyle(context: Context, readerStyle: Int) {
        runBlocking { ReaderPreferences.setReaderStyle(readerStyle) }
    }

    @JvmStatic
    fun getSavedTafsirKey(context: Context): String? {
        return ReaderPreferences.getTafsirId()
    }

    fun setSavedTafsirKey(context: Context, tafsirKey: String) {
        runBlocking { ReaderPreferences.setTafsirId(tafsirKey) }
    }
}
